docs/topics/db/instrumentation.txt

```diff
@@ -0,0 +1,118 @@
+========================
+Database instrumentation
+========================
+
+.. versionadded:: 2.0
+
+In order to help you understand and control the queries issued by your code,
+Django provides a mechanism for installing wrappers around the execution of
+database queries. Such wrappers can be used to count queries, measure the time
+they take, log them, or even prevent their execution (e.g. to make sure that no
+new database queries are issued while rendering a template, when you expect it
+to use only data which has been fetched already).
+
+The wrappers are modeled after :doc:`middleware </topics/http/middleware>` --
+they are callables which take, as one of their arguments, another callable;
+they call it to invoke the (possibly wrapped) database query, and they can do
+what they will around it. They are, however, created and installed by user
+code, and so do not need a separate factory like middleware do.
+
+Installing a wrapper is done in a context manager -- so the wrappers are
+temporary and can be made specific to some flow in your code.
+
+Probably the most trivial example of a wrapper is an execution blocker as
+mentioned above. It could look like this::
+
+    def blocker(*args):
+        raise Exception("No database access allowed here")
+
+And it would be used in a view to block queries from the template like so::
+
+    from django.db import connection
+    from django.shortcuts import render
+
+    def my_view(request):
+        context = ...  # Code to generate context with all data
+        template_name = ...
+        with connection.execute_wrapper(blocker):
+            return render(request, template_name, context)
+
+The parameters sent to the wrappers are:
+
+* ``execute`` -- a callable, which should be invoked with the rest of the
+  parameters in order to execute the query;
+
+* ``sql`` -- the SQL text to be sent to the database engine;
+
+* ``params`` -- parameters for the SQL statement (or statements);
+
+* ``many`` -- a Boolean indicating whether the ultimately invoked call is
+  ``execute()`` or ``executemany()`` (and whether ``params`` is expected to be
+  a list of values, or a list of lists of values);
+
+* ``context`` -- a dictionary with further data about the context of
+  invocation. For now, this includes the connection and cursor.
+
+So, a slightly more complex version of the blocker could include the connection
+name in the error message::
+
+    def blocker(execute, sql, params, many, context):
+        alias = context['connection'].alias
+        raise Exception("Access to database '{}' blocked here".format(alias))
+
+For a more complete example, a logger of queries could look like this::
+
+    import time
+
+    class QueryLogger:
+
+        def __init__(self):
+            self.queries = []
+
+        def log(execute, sql, params, many, context):
+            current_query = [sql, params, many]
```

Comment:
I think this would be easier to follow if `current_query` were a dict, then you would just add keys to it below, rather than all these `append` operations which are a bit hard to follow with the multiple execution paths and a full try/except/else/finally statement

Reply:
Good idea.